crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -124,12 +128,34 @@ impl Trade {
         })
     }
 
+    fn adjusted_order_limits(&self) -> Result<(eth::U256, eth::U256), Error> {
+        match self.policies.first() {
+            Some(order::FeePolicy::PriceImprovement {
+                factor: _,
+                max_volume_factor: _,
+                quote,
+            }) => adjust_quote_to_order_limits(
+                self.sell.amount.0,
+                self.buy.amount.0,
+                self.side,
+                quote.sell.amount.0,
+                quote.buy.amount.0,
+                quote.fee.amount.0,
+            )
+            .map_err(|e| e.into()),
+            _ => Ok((self.sell.amount.0, self.buy.amount.0)),
+        }
+    }
+
     /// Surplus based on custom clearing prices returns the surplus after all
     /// fees have been applied.
     ///
     /// Denominated in NATIVE token
     fn native_surplus(&self, prices: &auction::Prices) -> Result<eth::Ether, Error> {
-        let surplus = self.surplus().ok_or(Error::Surplus(self.sell, self.buy))?;
+        let (limit_sell_amount, limit_buy_amount) = self.adjusted_order_limits()?;
```

Comment:
I don't think we want to call `self.adjusted_order_limits()` here. Instead, we want to have:

`
let surplus = self
            .surplus_over_limit_price(self.sell.amount, self.buy.amount)
            .ok_or(Error::Surplus(self.sell, self.buy))?;
`

Reply:
But for in-market orders with a price improvement fee policy, that won't work since I assume native surplus should be compared against the quote in that case.

Reply:
The native surplus always works with order limit amounts.

Later, when we calculate the score, "protocol_fee" function will decide how much on top of the real surplus we want to add.

Reply:
I feel I did not explain this good enough. So, per CIP38, score = native_surplus + native_protocol_fee.

Native surplus in this case is a real surplus of the solution, as we have always defined it, a positive improvements over whatever user defined as limit amount (order amounts).